File: app/src/main/java/com/supercilex/robotscouter/ui/CardListHelper.java
package com.supercilex.robotscouter.ui;

import android.support.v7.widget.RecyclerView;

import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.supercilex.robotscouter.R;

public class CardListHelper {
    private final FirebaseRecyclerAdapter mAdapter;
    private final RecyclerView mRecyclerView;

    private final boolean mHasSafeCorners;

    public CardListHelper(FirebaseRecyclerAdapter adapter,
                          RecyclerView recyclerView,
                          boolean hasSafeCorners) {
        mAdapter = adapter;
        mRecyclerView = recyclerView;
        mHasSafeCorners = hasSafeCorners;
    }

    public void onBind(RecyclerView.ViewHolder viewHolder) {
        int position = viewHolder.getLayoutPosition();

        setBackground(viewHolder, position);


        if (mHasSafeCorners) return;

        // Update the items above and below to ensure the correct corner configuration is shown
        int abovePos = position - 1;
        int belowPos = position + 1;
        RecyclerView.ViewHolder above = mRecyclerView.findViewHolderForLayoutPosition(abovePos);
        RecyclerView.ViewHolder below = mRecyclerView.findViewHolderForLayoutPosition(belowPos);

        if (above != null) setBackground(above, abovePos);
        if (below != null) setBackground(below, belowPos);
    }

    protected boolean isFirstItem(int position) {
        return position == 0;
    }

    protected boolean isLastItem(int position) {
        return position == mAdapter.getItemCount() - 1;
    }

    private void setBackground(RecyclerView.ViewHolder viewHolder, int position) {
        int paddingLeft = viewHolder.itemView.getPaddingLeft();
        int paddingTop = viewHolder.itemView.getPaddingTop();
        int paddingRight = viewHolder.itemView.getPaddingRight();
        int paddingBottom = viewHolder.itemView.getPaddingBottom();

        boolean isFirstItem = isFirstItem(position);
        boolean isLastItem = isLastItem(position);

        if (isFirstItem && isLastItem) {
            viewHolder.itemView.setBackgroundResource(R.drawable.list_divider_single_item);
        } else if (isFirstItem) {
            viewHolder.itemView.setBackgroundResource(R.drawable.list_divider_first_item);
        } else if (isLastItem) {
            viewHolder.itemView.setBackgroundResource(R.drawable.list_divider_last_item);
        } else {
            viewHolder.itemView.setBackgroundResource(R.drawable.list_divider_middle_item);
        }

        viewHolder.itemView.setPadding(paddingLeft, paddingTop, paddingRight, paddingBottom);
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/CardListHelper.kt
package com.supercilex.robotscouter.ui

import android.support.v7.widget.RecyclerView
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.supercilex.robotscouter.R

open class CardListHelper(private val adapter: FirebaseRecyclerAdapter<*, *>,
                          private val recyclerView: RecyclerView,
                          private val hasSafeCorners: Boolean = false) {
    fun onBind(viewHolder: RecyclerView.ViewHolder) {
        val position: Int = viewHolder.layoutPosition

        setBackground(viewHolder, position)


        if (hasSafeCorners) return

        recyclerView.post {
            if (position > adapter.itemCount - 1) return@post

            // Update the items above and below to ensure the correct corner configuration is shown
            val abovePos = position - 1
            val belowPos = position + 1
            val above: RecyclerView.ViewHolder? = recyclerView.findViewHolderForLayoutPosition(abovePos)
            val below: RecyclerView.ViewHolder? = recyclerView.findViewHolderForLayoutPosition(belowPos)

            above?.let { setBackground(it, abovePos) }
            below?.let { setBackground(it, belowPos) }
        }
    }

    protected open fun isFirstItem(position: Int): Boolean {
        return position == 0
    }

    protected open fun isLastItem(position: Int): Boolean {
        return position == adapter.itemCount - 1
    }

    private fun setBackground(viewHolder: RecyclerView.ViewHolder, position: Int) {
        val itemView = viewHolder.itemView

        val paddingLeft = itemView.paddingLeft
        val paddingTop = itemView.paddingTop
        val paddingRight = itemView.paddingRight
        val paddingBottom = itemView.paddingBottom

        val isFirstItem = isFirstItem(position)
        val isLastItem = isLastItem(position)

        if (isFirstItem && isLastItem) {
            itemView.setBackgroundResource(R.drawable.list_divider_single_item)
        } else if (isFirstItem) {
            itemView.setBackgroundResource(R.drawable.list_divider_first_item)
        } else if (isLastItem) {
            itemView.setBackgroundResource(R.drawable.list_divider_last_item)
        } else {
            itemView.setBackgroundResource(R.drawable.list_divider_middle_item)
        }

        itemView.setPadding(paddingLeft, paddingTop, paddingRight, paddingBottom)
    }
}
